### PR TITLE
opentelemetry: support publishing metrics

### PR DESCRIPTION
**Pull request by bryangarza**
## Motivation
Currently, there is no way to publish metrics via tracing.

## Solution

[Rendered docs](https://deploy-preview-2185--tracing-rs.netlify.app/tracing_opentelemetry/struct.opentelemetrymetricssubscriber)

Update the tracing-opentelemetry crate to publish metrics for event fields that contain specific prefixes in the name.

Right now, we lazily instantiate and store one metrics object per-callsite, but a future improvement that we should add to tracing itself is the ability to store data per-callsite, so that we don't have to do a HashMap lookup each time we want to publish a metric.

## Example
It's this simple:
```rust
info!(MONOTONIC_COUNTER_API_REQUESTS = 1 as u64);
```

## Other notes
- Consider this a V1 that works end-to-end, but we will probably want follow-up PRs on top of this.
- If we had support for Valuable in the master branch, we could use an Enum instead of string prefixes (def. something we can follow up with)
- I did not support OTel observers ([docs](https://docs.rs/opentelemetry/latest/opentelemetry/metrics/struct.Meter.html)), since the same metrics events should be consumable by other subscribers such as tokio-console, they should not be specific to OTel. I'm thinking if someone wants to use observers, they should directly use the rust-opentelemetry crate in their code for that (which might require changing the code slightly so that the application has access to the `Meter`)
- Right now this is only for Events, but it should take very little additional work to also support Spans (follow-up PR?)
- I have a separate repo where I was testing this out, you can find it here: https://github.com/bryangarza/tracing-metrics-feature-test

### CODE REVIEW

File: tracing-opentelemetry/src/subscriber.rs [hunk 1]
```diff
@@ -21,14 +29,23 @@ const SPAN_NAME_FIELD: &str = "otel.name";
 const SPAN_KIND_FIELD: &str = "otel.kind";
 const SPAN_STATUS_CODE_FIELD: &str = "otel.status_code";
 const SPAN_STATUS_MESSAGE_FIELD: &str = "otel.status_message";
+const INSTRUMENTATION_NAME: &str = "tracing/tracing-opentelemetry";
+
+struct Instruments_(Arc<RwLock<Instruments>>);
 
+struct MetricsHandler {
+    _push_controller: PushController,
```
Comment (bryangarza) [inline tracing-opentelemetry/src/subscriber.rs:37]: I put underscore because it was unused right now (only needed to create the `Meter`), but I think we might want to require this to be passed in, so that we can make use of it later on in follow-up improvements 🤔

**Comment by jtescher** [comment]
Great to get started thinking about how otel-metrics fits in to tracing, as the metrics impl is still in progress you may have an easier time maintaining a separate subscriber/layer that you can update without worrying as much about breaking changes / stability guarantees. e.g. https://github.com/open-telemetry/opentelemetry-rust/pull/819 is a fairly substantial change to the otel metrics api and configuring metric views / multiple metric exporters is still to be built out.

Other option could be a feature flag that suggests the instability / warns about using it currently, but it may be more trouble than it's worth until the metrics side stablizes a bit more.

**Comment by bryangarza** [comment]
I ran into an issue with my integration tests where if I put the assertions into the test function, the assertions seem to run *before* `MetricsExporter::export`, making it impossible to pass values from its closure into some shared state that the test function can check. As far as I know though, `tokio::test` runs single-threaded by default. Do I need to do something to force the `MetricsExporter::export` method to be called eagerly? I'm guessing that `rust-opentelemetry` is waiting for some buffer to fill before flushing the metrics out?

File: tracing-opentelemetry/src/metrics.rs [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+        };
+    }
+}
+
+pub(crate) struct MetricVisitor<'a> {
+    pub(crate) instruments: &'a Arc<RwLock<Instruments>>,
+    pub(crate) meter: &'a Meter,
+}
+
+// impl<'a> Visit for MetricVisitor<'a> {
```
Comment (bryangarza) [inline tracing-opentelemetry/src/metrics.rs:110]: Need to delete this line, I'll do it in my next commit

**Comment by bryangarza** [comment]
This morning I tried to improve the tests again, but the `CheckpointSet` behavior is not consistent, sometimes, when `MetricsExporter::export` is called there is data available, sometimes (the majority of the time) the `CheckpointSet` is empty. I have a separate branch in case anyone is curious, and there's a bit more explanation in the commit message https://github.com/bryangarza/tracing/commit/4804dcef2df957de0e170f86fc417406e5a1d392

Other than that, I think this is ready to merge @carllerche

**Comment by bryangarza** [comment]
Got some good feedback from @hlbarber (thanks!), which I've now incorporated into the PR (using `&'static str` in the `MetricVisitor` to reduce allocations)

File: tracing-opentelemetry/src/metrics.rs [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+/// first time that a metric is emitted by the instrumentation, a `Metric`
+/// instance will be created and added to the corresponding map. This means that
+/// any time a metric is emitted by the instrumentation, one map lookup has to
+/// be performed.
+///
+/// In the future, this can be improved by associating each Metric instance to
+/// its callsite. However, per-callsite storage is not yet supported by tracing.
+pub struct OpenTelemetryMetricsSubscriber {
+    meter: Meter,
+    instruments: Arc<RwLock<Instruments>>,
```
Comment (hawkw) [inline tracing-opentelemetry/src/metrics.rs:305]: it shouldn't be necessary to place a single global `RwLock` around *all* the different metric types. updating a metric value or creating a new metric will only mutate one of the `HashMap`s in the `Instruments` struct at a time. however, by placing the whole thing inside of one big lock, this forces *all* metric updates to be globally synchronized, even for different types of metric. in the current implementation, updating a `u64` value recorder metric (for example) will prevent another thread from recording a `f64` up-down counter at the same time.

i think we should make the scope of the `RwLock`s much more granular.

**Review by hawkw — CHANGES_REQUESTED** [review]
i noticed some potential performance issues with the current implementation that seem fairly significant. i've opened a PR against this branch (https://github.com/bryangarza/tracing/pull/1) that rewrites some of this code to reduce the overhead of string allocation and global locking.

File: tracing-opentelemetry/src/metrics.rs [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+    CounterF64(f64),
+    UpDownCounterI64(i64),
+    UpDownCounterF64(f64),
+    ValueRecorderU64(u64),
+    ValueRecorderI64(i64),
+    ValueRecorderF64(f64),
+}
+
+impl Instruments {
+    pub(crate) fn init_metric_for(
```
[the remark this answers is not in the document]
Comment (bryangarza) [reply tracing-opentelemetry/src/metrics.rs]: yeah you're right the name is definitely a minomer

File: tracing-opentelemetry/src/metrics.rs [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+        &mut self,
+        meter: &Meter,
+        instrument_type: InstrumentType,
+        metric_name: String,
```
[the remark this answers is not in the document]
Comment (bryangarza) [reply tracing-opentelemetry/src/metrics.rs]: Gotcha, yeah I had changed this to be `&'static str` a bit ago, I had forgotten that the metadata name field is static

File: tracing-opentelemetry/src/metrics.rs (hunk 3 above)
Comment (bryangarza) [reply tracing-opentelemetry/src/metrics.rs]: Definitely agree.

**Comment by bryangarza** [comment]
I cancelled the 5 CI tests that were running because seems that they are now hanging. Tried this out locally as well, but need to dig into it further.

File: tracing-opentelemetry/src/metrics.rs [hunk 6]
```diff
@@ -53,10 +53,12 @@ impl Instruments {
...
-            if let Some(metric) = lock.get(name) {
-                update(metric);
-                return;
+            {
+                let lock = map.read().unwrap();
+                if let Some(metric) = lock.get(name) {
+                    update(metric);
+                    return;
+                }
             }
```
[the remark this answers is not in the document]
Comment (bryangarza) [reply tracing-opentelemetry/src/metrics.rs:62]: thanks!
yeah i just recently learned that when you reuse a variable name, rust doesn't drop the previous usage in case something else depends on that data

**Review by carllerche — APPROVED** [review]
LGTM, thanks

**Review by hawkw — CHANGES_REQUESTED** [review]
I have basically one remaining concern about this PR, which is that some aspects of the current implementation of the metrics code make it more or less impossible to implement metrics that follow [OpenTelemetry's metric naming conventions](https://opentelemetry.io/docs/reference/specification/metrics/semantic_conventions/)

In particular, the metric names in this PR are
 - required to start with a prefix, which is part of the name of the metric that's output to OpenTelemetry
 - uppercase, for some reason?

I think that using prefixes to indicate the type of the metric is fine for now (although I'm definitely open to nicer ways of specifying metric types in the future), but I think we should change the code for recognizing metrics based on field name prefixes so that the metric value that's actually emitted to the OpenTelemetry collector doesn't include a prefix like `MONOTONIC_COUNTER` in the actual name of the metric. We could do this by using [`str::strip_prefix`](https://doc.rust-lang.org/stable/std/primitive.str.html#method.strip_prefix) in the visitor implementation, rather than `starts_with`, and using the field name without the prefix as the hashmap key.

In addition, I think we may want to change the prefixes we recognize. Both `tracing` and OpenTelemetry support dots in field names, and semantically, a dot can be used as a namespacing operator. I think we should probably use prefixes like `monotonic_counter.` rather than `MONOTONIC_COUNTER_` to more clearly separate the metric type from the rest of the metric name.

So, in summary, I think we should:
 - [x] make the metric name prefixes lowercase
 - [x] make the metric name prefixes end in a `.`
 - [x] strip the prefix when matching metric names before using them as hashmap keys
 - [ ] change the metric names in the various examples and tests to be all lowercase

and then I'll be very happy to merge this PR!

**Comment by bryangarza** [comment]
> * [x] make the metric name prefixes lowercase
> 
> * [x]  make the metric name prefixes end in a `.`
> 
> * [x]  strip the prefix when matching metric names before using them as hashmap keys
> 
> * [x]  change the metric names in the various examples and tests to be all lowercase

@eliza, thanks -- done in my last commit

File: tracing-opentelemetry/src/metrics.rs [hunk 7]
```diff
@@ -0,0 +1,360 @@
...
+///
+/// This subscriber holds a set of maps, each map corresponding to a type of
+/// metric supported by OpenTelemetry. These maps are populated lazily. The
+/// first time that a metric is emitted by the instrumentation, a `Metric`
+/// instance will be created and added to the corresponding map. This means that
+/// any time a metric is emitted by the instrumentation, one map lookup has to
+/// be performed.
+///
+/// In the future, this can be improved by associating each Metric instance to
+/// its callsite. However, per-callsite storage is not yet supported by tracing.
```
[the remark this answers is not in the document]
Comment (bryangarza) [reply tracing-opentelemetry/src/metrics.rs:330]: Why remove this line? `However, per-callsite storage is not yet supported by tracing.`? @davidbarsky curious the reason

File: tracing-opentelemetry/src/metrics.rs [hunk 8]
```diff
@@ -0,0 +1,360 @@
...
+    meter: Meter,
+    instruments: Instruments,
+}
+
+impl OpenTelemetryMetricsSubscriber {
+    /// Create a new instance of OpenTelemetryMetricsSubscriber.
+    pub fn new(push_controller: PushController) -> Self {
+        let meter = push_controller
+            .provider()
+            .meter(INSTRUMENTATION_LIBRARY_NAME, Some(CARGO_PKG_VERSION));
```
[the remark this answers is not in the document]
Comment (bryangarza) [reply tracing-opentelemetry/src/metrics.rs:341]: yeah it's something that opentelemetry includes during metrics exporting

File: tracing-opentelemetry/src/metrics.rs [hunk 9]
```diff
@@ -0,0 +1,360 @@
...
+/// # let push_controller: PushController = unimplemented!();
+///
+/// let opentelemetry_metrics =  OpenTelemetryMetricsSubscriber::new(push_controller);
+/// let collector = Registry::default().with(opentelemetry_metrics);
+/// tracing::collect::set_global_default(collector).unwrap();
+/// ```
+///
+/// To publish a new metric from your instrumentation point, all that is needed
+/// is to add a key-value pair to your `tracing::Event` that contains a specific
+/// prefix. One of the following:
```
[the remark this answers is not in the document]
Comment (bryangarza) [reply tracing-opentelemetry/src/metrics.rs:249]: I'll update this is a new commit, cause seems that the suggestion has an extra word `from`

File: tracing-opentelemetry/src/metrics.rs [hunk 10]
```diff
@@ -0,0 +1,360 @@
...
+/// This subscriber holds a set of maps, each map corresponding to a type of
+/// metric supported by OpenTelemetry. These maps are populated lazily. The
+/// first time that a metric is emitted by the instrumentation, a `Metric`
+/// instance will be created and added to the corresponding map. This means that
+/// any time a metric is emitted by the instrumentation, one map lookup has to
+/// be performed.
+///
+/// In the future, this can be improved by associating each Metric instance to
+/// its callsite. However, per-callsite storage is not yet supported by tracing.
+pub struct OpenTelemetryMetricsSubscriber {
```
[the remark this answers is not in the document]
Comment (bryangarza) [reply tracing-opentelemetry/src/metrics.rs:331]: ok

[the remark this answers is not in the document]
Comment (bryangarza) [reply tracing-opentelemetry/src/metrics.rs]: done

File: tracing-opentelemetry/src/metrics.rs [hunk 11]
```diff
@@ -0,0 +1,360 @@
...
+/// Do not mix floating-point and non-floating-point numbers for the same
+/// metric. Instead, if you are ever going to use floating-point numbers for a
+/// metric, cast any other usages of that metric to `f64`.
+///
+/// Don't do this:
+/// ```
+/// # use tracing::info;
+/// info!(MONOTONIC_COUNTER_FOO = 1);
+/// info!(MONOTONIC_COUNTER_FOO = 1.0);
+/// ```
```
[the remark this answers is not in the document]
Comment (bryangarza) [reply tracing-opentelemetry/src/metrics.rs]: will remove

[the remark this answers is not in the document]
Comment (bryangarza) [reply tracing-opentelemetry/src/metrics.rs]: done

File: tracing-opentelemetry/src/metrics.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (bryangarza) [reply tracing-opentelemetry/src/metrics.rs]: done

File: tracing-opentelemetry/src/metrics.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (bryangarza) [reply tracing-opentelemetry/src/metrics.rs:322]: Anyways, it's fine. I can remove it. Just need to get this merged soon.

[the remark this answers is not in the document]
Comment (bryangarza) [reply tracing-opentelemetry/src/metrics.rs]: Done
